### PR TITLE
UGENE-7785 [Pcr, crash] Cancel PCR workflow

### PR DESCRIPTION
**Pull request by DmitriiSukhomlinov**

### CODE REVIEW

File: src/plugins/pcr/src/InSilicoPcrWorker.cpp [hunk 1]
```diff
@@ -192,6 +192,7 @@ Task* InSilicoPcrWorker::createPrepareTask(U2OpStatus& os) const {
     if (task == nullptr) {
         os.setError(tr("Can not read the primers file: ") + primersUrl);
     }
+    task->moveDocumentToMainThread = true;
```
Comment (DmitriiSukhomlinov) [inline src/plugins/pcr/src/InSilicoPcrWorker.cpp:195]: Not really related to the crash, but resolves [this](https://ugene.dev/tracker/browse/UGENE-7785?focusedCommentId=31988&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-31988) unpredictable behaviour

File: src/plugins/GUITestBase/src/GTUtilsTaskTreeView.cpp [hunk 2]
```diff
@@ -236,11 +252,31 @@ int GTUtilsTaskTreeView::countTasks(HI::GUITestOpStatus& os, const QString& item
...
     GTGlobals::sleep(500);
     return getTreeWidgetItem(os, itemName)->text(1);
 }
+#undef GT_METHOD_NAME
+
+#define GT_METHOD_NAME "getTaskProgress"
+int GTUtilsTaskTreeView::getTaskProgress(HI::GUITestOpStatus& os, const QString& itemName, bool failOnNull) {
+    if (!isViewOpened(os)) {
+        openView(os);
+        GTGlobals::sleep(500);
```
Comment (yalgaer) [inline src/plugins/GUITestBase/src/GTUtilsTaskTreeView.cpp]: Do not sleep a constant time: it is too slow and may be not enough. Use findWidget()

File: src/plugins/GUITestBase/src/GTUtilsTaskTreeView.h [hunk 3]
```diff
@@ -75,6 +80,7 @@ class GTUtilsTaskTreeView {
 
     static int countTasks(HI::GUITestOpStatus& os, const QString& itemName);
     static QString getTaskStatus(HI::GUITestOpStatus& os, const QString& itemName);
+    static int getTaskProgress(HI::GUITestOpStatus& os, const QString& itemName, bool failOnNull = true);
```
Comment (yalgaer) [inline src/plugins/GUITestBase/src/GTUtilsTaskTreeView.h]: Please document what kind of ints (like -1) the method returns. 
Rename `failOnNull` to `findIfNotFound`

File: src/plugins/GUITestBase/src/GTUtilsTaskTreeView.cpp [hunk 4]
```diff
@@ -236,11 +252,31 @@ int GTUtilsTaskTreeView::countTasks(HI::GUITestOpStatus& os, const QString& item
...
+    }
+    auto item = getTreeWidgetItem(os, itemName, failOnNull);
+    CHECK(item != nullptr, -1);
+
+    auto itemText = item->text(2);
+    bool ok = false;
+    auto number = itemText.left(2).toInt(&ok);
```
Comment (yalgaer) [inline src/plugins/GUITestBase/src/GTUtilsTaskTreeView.cpp]: What if the progress is `1%`? Will `left(2)` work in this case?

File: src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp [hunk 5]
```diff
@@ -4462,5 +4463,51 @@ GUI_TEST_CLASS_DEFINITION(test_7863) {
...
+    auto stopButton = GTAction::button(os, "Stop workflow", GTUtilsMdi::activeWindow(os));
+    auto globalPos = stopButton->mapToGlobal(GTWidget::getWidgetVisibleCenter(stopButton));
+    GTMouseDriver::moveTo(globalPos);
+
+    GTUtilsTask::waitTaskStart(os, "Multiple In Silico PCR");
+
+    int progress;
+    do {
+        progress = GTUtilsTaskTreeView::getTaskProgress(os, "Multiple In Silico PCR", false);
+    } while (progress < 91);
```
Comment (yalgaer) [inline src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp]: Do not run an infinite loop with no sleep() inside. 
Use a pattern like
```
for (int time = 0; time < GT_OP_WAIT_MILLIS && progress < 91; time += GT_OP_CHECK_MILLIS) {
        GTGlobals::sleep(time > 0 ? GT_OP_CHECK_MILLIS : 0);
...
}
```

File: src/plugins/pcr/src/InSilicoPcrWorker.cpp (hunk 1 above)
Comment (yalgaer) [reply src/plugins/pcr/src/InSilicoPcrWorker.cpp:195]: ok

File: src/plugins/pcr/src/InSilicoPcrWorkflowTask.cpp [hunk 6]
```diff
@@ -37,7 +39,9 @@ QList<Task*> InSilicoPcrWorkflowTask::onSubTaskFinished(Task* subTask) {
 
     if (subTask == pcrTask) {
         foreach (const InSilicoPcrProduct& product, pcrTask->getResults()) {
-            auto productTask = new ExtractProductTask(product, productSettings);
+            QVariantMap hints;
+            hints[DocumentRemovalMode_Synchronous] = QString();
```
Comment (yalgaer) [inline src/plugins/pcr/src/InSilicoPcrWorkflowTask.cpp]: nit: use `true`, not `QString()` like for other boolean-like hints

**Review by yalgaer — CHANGES_REQUESTED** [review]
See comments

File: src/plugins/GUITestBase/src/GTUtilsTaskTreeView.cpp (hunk 2 above)
Comment (DmitriiSukhomlinov) [reply src/plugins/GUITestBase/src/GTUtilsTaskTreeView.cpp]: I replaced it from the similar function above `GTUtilsTaskTreeView::getTaskStatus()`. Not sure, what is it for, deletion doesn't affect anything.

File: src/plugins/GUITestBase/src/GTUtilsTaskTreeView.h (hunk 3 above)
Comment (DmitriiSukhomlinov) [reply src/plugins/GUITestBase/src/GTUtilsTaskTreeView.h]: OK

File: src/plugins/GUITestBase/src/GTUtilsTaskTreeView.cpp (hunk 4 above)
Comment (DmitriiSukhomlinov) [reply src/plugins/GUITestBase/src/GTUtilsTaskTreeView.cpp]: OK

File: src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp (hunk 5 above)
Comment (DmitriiSukhomlinov) [reply src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp]: OK, good

File: src/plugins/pcr/src/InSilicoPcrWorkflowTask.cpp (hunk 6 above)
Comment (DmitriiSukhomlinov) [reply src/plugins/pcr/src/InSilicoPcrWorkflowTask.cpp]: OK

File: src/plugins/GUITestBase/src/GTUtilsTaskTreeView.h [hunk 7]
```diff
@@ -75,6 +80,12 @@ class GTUtilsTaskTreeView {
 
     static int countTasks(HI::GUITestOpStatus& os, const QString& itemName);
     static QString getTaskStatus(HI::GUITestOpStatus& os, const QString& itemName);
+    /*f
```
Comment (yalgaer) [inline src/plugins/GUITestBase/src/GTUtilsTaskTreeView.h]: nit: 
1) remove 'f'
2) Rename 'findIfNotFound' to 'failIfNotFound'

File: src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp [hunk 8]
```diff
@@ -4462,5 +4463,52 @@ GUI_TEST_CLASS_DEFINITION(test_7863) {
...
+    GTUtilsTaskTreeView::doubleClick(os, "Execute workflow");
+    GTUtilsTaskTreeView::doubleClick(os, "Workflow run");
+
+    auto stopButton = GTAction::button(os, "Stop workflow", GTUtilsMdi::activeWindow(os));
+    auto globalPos = stopButton->mapToGlobal(GTWidget::getWidgetVisibleCenter(stopButton));
+    GTMouseDriver::moveTo(globalPos);
+
+    GTUtilsTask::waitTaskStart(os, "Multiple In Silico PCR");
+
+    int progress;
```
Comment (yalgaer) [inline src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp]: This variable must be initialized before use. I would recommend moving this method into GTUtilsTaskTreeView as an utility method

**Review by yalgaer — CHANGES_REQUESTED** [review]
Looks good except 2 minor issues. Please update/run tests and I will approve.

File: src/plugins/GUITestBase/src/GTUtilsTaskTreeView.h (hunk 7 above)
Comment (DmitriiSukhomlinov) [reply src/plugins/GUITestBase/src/GTUtilsTaskTreeView.h]: OK

File: src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp (hunk 8 above)
Comment (DmitriiSukhomlinov) [reply src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp]: OK

**Review by yalgaer — COMMENTED** [review]
Looks good. Please remove unrelated changes from the patch (use git merge to avoid issues like this) and I will approve after tests.

**Review by yalgaer — APPROVED** [review]

**Pull request merged**